gossipsub_test.go

```diff
@@ -1717,3 +1717,305 @@ func (ps *mockPeerScoreInspector) score(p peer.ID) float64 {
 	defer ps.mx.Unlock()
 	return ps.scores[p]
 }
+
+func TestGossipsubRPCFragmentation(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	hosts := getNetHosts(t, ctx, 2)
+	ps := getGossipsub(ctx, hosts[0])
+
+	// make a fake peer that requests everything through IWANT gossip
+	iwe := iwantEverything{h: hosts[1]}
+	iwe.h.SetStreamHandler(GossipSubID_v10, iwe.handleStream)
+
+	connect(t, hosts[0], hosts[1])
+
+	// have the real pubsub join the test topic
+	_, err := ps.Subscribe("test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait for the real pubsub to connect and try to graft to the faker
+	time.Sleep(time.Second)
+
+	// publish a bunch of fairly large messages from the real host
+	nMessages := 1000
+	msgSize := 20000
+	for i := 0; i < nMessages; i++ {
+		msg := make([]byte, msgSize)
+		rand.Read(msg)
+		ps.Publish("test", msg)
+		time.Sleep(20 * time.Millisecond)
+	}
+
+	// wait a bit for them to be received via gossip by the fake peer
+	time.Sleep(5 * time.Second)
+	iwe.lk.Lock()
+	defer iwe.lk.Unlock()
+
+	// we should have received all the messages
+	if iwe.msgsReceived != nMessages {
+		t.Fatalf("expected fake gossipsub peer to receive all messages, got %d / %d", iwe.msgsReceived, nMessages)
+	}
+
+	// and we should have seen an IHAVE message for each of them
+	if iwe.ihavesReceived != nMessages {
+		t.Fatalf("expected to get IHAVEs for every message, got %d / %d", iwe.ihavesReceived, nMessages)
+	}
+
+	// If everything were fragmented with maximum efficiency, we would expect to get
+	// (nMessages * msgSize) / ps.maxMessageSize total RPCs containing the messages we sent IWANTs for.
+	// The actual number will probably be larger, since there's some overhead for the RPC itself, and
+	// we probably aren't packing each RPC to it's maximum size
+	minExpectedRPCS := (nMessages * msgSize) / ps.maxMessageSize
+	if iwe.rpcsWithMessages < minExpectedRPCS {
+		t.Fatalf("expected to receive at least %d RPCs containing messages, got %d", minExpectedRPCS, iwe.rpcsWithMessages)
+	}
+}
+
+// iwantEverything is a simple gossipsub client that never grafts onto a mesh,
+// instead requesting everything through IWANT gossip messages. It is used to
+// test that large responses to IWANT requests are fragmented into multiple RPCs.
+type iwantEverything struct {
+	h                host.Host
+	lk               sync.Mutex
+	rpcsWithMessages int
+	msgsReceived     int
+	ihavesReceived   int
+}
+
+func (iwe *iwantEverything) handleStream(s network.Stream) {
+	os, err := iwe.h.NewStream(context.Background(), s.Conn().RemotePeer(), GossipSubID_v10)
```

Comment:
we should probably add a `defer s.Close()` here (not that it matters much, but let's be nice)